src/multicore.rs

```diff
@@ -48,10 +54,51 @@ impl Worker {
         R: Send + 'static,
     {
         let (sender, receiver) = bounded(1);
-        THREAD_POOL.spawn(move || {
-            let res = f();
-            sender.send(res).unwrap();
-        });
+
+        let thread_index = if THREAD_POOL.current_thread_index().is_some() {
+            THREAD_POOL.current_thread_index().unwrap()
+        } else {
+            0
+        };
+
+        // We keep track here of how many times spawn has been called.
+        // It can be called without limit, each time, putting a
+        // request for a new thread to execute a method on the
+        // ThreadPool.  However, if we allow it to be called without
+        // limits, we run the risk of memory exhaustion due to limited
+        // stack space consumed by all of the pending closures to be
+        // executed.
+        let previous_count = WORKER_SPAWN_COUNTER.fetch_add(1, Ordering::SeqCst);
+
+        // If the number of spawns requested has exceeded the number
+        // of cores available for processing by some factor (the
+        // default being 4), instead of requesting that we spawn a new
+        // thread, we instead execute the closure in the context of an
+        // install call to help clear the growing work queue and
+        // minimize the chances of memory exhaustion.
+        if previous_count > *WORKER_SPAWN_MAX_COUNT {
+            THREAD_POOL.install(move || {
+                debug!("[{}] switching to install to help clear backlog[current threads {}, threads requested {}]",
```

Comment:
Perhaps this should even be `trace!()` as I'd expect it to create quite a lot of noise.